### ISSUE TITLE
Make a tutorial for porting PyTorch models to candle

### ISSUE THREAD

**Issue by LaurentMazare** [huggingface/candle#1075]
Most of the candle models have been ported from the Python `transformers` library (you can often find in the comments pointers to the source for the Python implementation so that it's easy to look at the two versions at the same time). It would be great to document the process of "porting" models so as to get new users/contributors started.

For me the process consists in looking at the Python source in `transformers` nearly from top to bottom and converting the Python classes to equivalent Rust structs. I usually create a small binary crate to hold the code so as to avoid having to recompile everything each time.

The PyTorch and candle apis are pretty similar so that it's usually not that hard to do. Here is a sample class [Python implementation](https://github.com/huggingface/transformers/blob/e1cec43415e72c9853288d4e9325b734d36dd617/src/transformers/models/t5/modeling_t5.py#L280) and the [Rust equivalent](https://github.com/huggingface/candle/blob/e7560443e4680b7655d011948d3cf178268fcfff/candle-transformers/src/models/t5.rs#L161-L182) - note that the tracing bits are not really necessary.

Once that part is done, I add some model calls to the binary to perform text generation/image recognition/... - this is usually just copied from another example.

Then I try to run the model with some `.safetensor` loaded weights - if only PyTorch weights are available I would write a tiny Python script to export them to `.safetensor`.
```python
import torch
from safetensors import safe_open
from safetensors.torch import save_file

model = torch.load("convmixer_1024_20_ks9_p14.pth.tar", map_location=torch.device('cpu'))
save_file(model, "convmixer_1024_20_ks9_p14.safetensors")
```

At this point if I run into shape mismatches, I would run with `RUST_BACKTRACE=1` to find the culprit, these are usually straightforward to correct.

Finally if the predictions match the PyTorch one it's all good, if not I would print the tensors on the candle and PyTorch side for the first block/layer so as to find out the first point where they differ.

**Comment by kingzcheung** [comment]
Great experience sharing. I'm also trying to port some early models, such as ResNet. My previous confusion was that I wasn't sure if Candle could fully support PyTorch models.

**Comment by ToluClassics** [comment]
I'd love to help with this @LaurentMazare  if there are opportunities to!. I've been porting some dense retrievers to rust mostly using tch-rs. With some guidance, I can get started on this

**Comment by LaurentMazare** [comment]
@ToluClassics that would be great, what I would suggest is port a model that you care about and find useful. Better if the model is simple, has a PyTorch implementation (usually even nicer if it's in the transformer library the code is pretty easy to adapt), has some pre-trained weights, and ideally is of general usefulness (we've ported mostly very recent models so far but older models are good too). Then make a new repo to host the model and document the process of converting it, I think it would be good to have some narrative around a simple Python class that gets converted to an equivalent candle struct (and so the `__init__` in python becomes the `new` in rust, the `__call__` becomes `forward` and the method themselves get adapted). Maybe give this a try and if you run into any difficulties ping me (just tag me in an issue on your repo).

**Comment by lunarflu** [comment]
cool issue, x-posting this to the HF discord! 🕯️ 🚀

**Comment by kingzcheung** [comment]
@LaurentMazare  Whether caddle has a MaxPool2d equivalent implementation？
```
nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
```
It (`max_pool2d_with_stride(3, 2)` ) does not work equivalently.

**Comment by testdepth** [comment]
would porting `pipelines` in a similar manner be a useful exercise? seems to be mostly unimplemented in the main branch

**Comment by LaurentMazare** [comment]
@kingzcheung `max_pool2d_with_stride` doesn't apply any padding so you would have to add it on your side before applying the operator (using pad-with-zeros or something alike). Besides this, the implementation seems to match pytorch as can be tried out via the following script:
```python
import torch
import candle

t = torch.randn((10, 8))
print(torch.nn.functional.max_pool2d(t.unsqueeze(0).unsqueeze(0), 3, stride=2))

t2 = candle.Tensor(t.flatten().tolist()).reshape((1, 1, 10, 8))
print(candle.functional.max_pool2d(t2, 3, stride=2))

print(torch.nn.functional.avg_pool2d(t.unsqueeze(0).unsqueeze(0), 3, stride=2))

t2 = candle.Tensor(t.flatten().tolist()).reshape((1, 1, 10, 8))
print(candle.functional.avg_pool2d(t2, 3, stride=2))
```

**Comment by ToluClassics** [comment]
> would porting `pipelines` in a similar manner be a useful exercise? seems to be mostly unimplemented in the main branch

Pinging @LaurentMazare on this again

**Comment by kingzcheung** [comment]
@LaurentMazare  I want to know whether the following writing is equivalent in both tch and caddle. When I locate the problem, I keep getting different tensor results？

model:
```
(conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
```
tch:
```
let conv1 = conv2d(&p / "conv1", 64, 64, 3, 1, 1);
```
caddle:
```
let conv_config = nn::Conv2dConfig {1,1,..Default::default()}
nn::conv2d_no_bias(64, 64, 3, conv_config, vb)
```

**Comment by LaurentMazare** [comment]
> caddle:
> 
> ```
> let conv_config = nn::Conv2dConfig {1,1,..Default::default()}
> nn::conv2d_no_bias(64, 64, 3, conv_config, vb)
> ```

That would be roughly it. Better to name the fields appropriately in the conv config, the default stride is already 1 so no need to specify it.
```rust
let conv_config = nn::Conv2dConfig { padding: 1, ..Default::default()};
nn::conv2d_no_bias(64, 64, 3, conv_config, vb.pp("conv1"))
```

**Comment by kingzcheung** [comment]
> @LaurentMazare I want to know whether the following writing is equivalent in both tch and caddle. When I locate the problem, I keep getting different tensor results？
> 
> model:
> 
> ```
> (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
> ```
> 
> tch:
> 
> ```
> let conv1 = conv2d(&p / "conv1", 64, 64, 3, 1, 1);
> ```
> 
> caddle:
> 
> ```
> let conv_config = nn::Conv2dConfig {1,1,..Default::default()}
> nn::conv2d_no_bias(64, 64, 3, conv_config, vb)
> ```
> 
> I write resnet in candle.
> 
> ```
> use candle_core::{ Result, D };
> use candle_nn as nn;
> use nn::{ Module, VarBuilder, Conv2d, Linear, BatchNormConfig };
> 
> #[derive(Debug)]
> pub struct Sequential<T: Module> {
>     layers: Vec<T>,
> }
> 
> pub fn seq<T: Module>() -> Sequential<T> {
>     Sequential { layers: vec![] }
> }
> 
> impl<T: Module> Sequential<T> {
>     pub fn len(&self) -> usize {
>         self.layers.len()
>     }
> 
>     pub fn is_empty(&self) -> bool {
>         self.layers.is_empty()
>     }
> 
>     pub fn add(&mut self, layer: T) {
>         self.layers.push(layer);
>     }
> }
> 
> impl<T: Module> Module for Sequential<T> {
>     fn forward(&self, xs: &candle_core::Tensor) -> Result<candle_core::Tensor> {
>         if self.layers.is_empty() {
>             Ok(xs.clone())
>         } else {
>             let xs = self.layers[0].forward(xs)?;
>             println!("base_block:{xs}");
>             self.layers
>                 .iter()
>                 .skip(1)
>                 .try_fold(xs, |xs, layer| layer.forward(&xs))
>         }
>     }
> }
> 
> /// 1x1 convolution
> fn conv2d(
>     vb: VarBuilder,
>     in_planes: usize,
>     out_planes: usize,
>     ksize: usize,
>     padding: usize,
>     stride: usize,
>     bias: bool
> ) -> Result<nn::Conv2d> {
>     let conv_config = nn::Conv2dConfig {
>         stride,
>         padding,
>         ..Default::default()
>     };
>     if bias {
>         nn::conv2d(in_planes, out_planes, ksize, conv_config, vb)
>     } else {
>         nn::conv2d_no_bias(in_planes, out_planes, ksize, conv_config, vb)
>     }
> }
> 
> #[derive(Debug)]
> pub struct Downsample {
>     conv2d: nn::Conv2d,
>     bn2: nn::BatchNorm,
> }
> 
> impl Downsample {
>     fn new(in_planes: usize, out_planes: usize, stride: usize, vb: VarBuilder) -> Result<Self> {
>         let conv2d = conv2d(vb.pp("0"), in_planes, out_planes, 1, 0, stride, false)?;
> 
>         let config = BatchNormConfig { eps: 1e-5, remove_mean: false, affine: true };
>         let bn2 = nn::batch_norm(out_planes, config, vb.pp("1"))?;
>         Ok(Self { conv2d, bn2 })
>     }
> }
> 
> impl Module for Downsample {
>     fn forward(&self, xs: &candle_core::Tensor) -> Result<candle_core::Tensor> {
>         let xs = self.conv2d.forward(xs)?;
>         let xs = self.bn2.forward(&xs)?;
>         Ok(xs)
>     }
> }
> 
> fn downsample(
>     vb: VarBuilder,
>     in_planes: usize,
>     out_planes: usize,
>     stride: usize
> ) -> Result<Option<Downsample>> {
>     if stride != 1 || in_planes != out_planes {
>         Ok(Some(Downsample::new(in_planes, out_planes, stride, vb)?))
>     } else {
>         Ok(None)
>     }
> }
> 
> #[derive(Debug)]
> pub struct BasicBlock {
>     conv1: nn::Conv2d,
>     bn1: nn::BatchNorm,
>     conv2: nn::Conv2d,
>     bn2: nn::BatchNorm,
>     downsample: Option<Downsample>,
> }
> 
> impl BasicBlock {
>     
>     pub fn new(vb: VarBuilder, in_planes: usize, out_planes: usize, stride: usize) -> Result<Self> {
>         let conv1 = conv2d(vb.pp("conv1"), in_planes, out_planes, 3, 1, stride, false)?;
>         let bn1 = nn::batch_norm(
>             out_planes,
>             BatchNormConfig { eps: 1e-5, remove_mean: false, affine: true },
>             vb.pp("bn1")
>         )?;
>         let conv2 = conv2d(vb.pp("conv2"), out_planes, out_planes, 3, 1, 1, false)?;
>         let bn2 = nn::batch_norm(
>             out_planes,
>             BatchNormConfig { eps: 1e-5, remove_mean: false, affine: true },
>             vb.pp("bn2")
>         )?;
>         let downsample = downsample(vb.pp("downsample"), in_planes, out_planes, stride)?;
> 
>         Ok(Self { conv1, bn1, conv2, bn2, downsample })
>     }
> }
> 
> impl Module for BasicBlock {
>     fn forward(&self, xs: &candle_core::Tensor) -> Result<candle_core::Tensor> {
>         let ys = xs.apply(&self.conv1)?;
>         // println!("conv1:: {ys}");
>         let ys = ys.apply(&self.bn1)?;
>         let ys = ys.relu()?;
>         let ys = ys.apply(&self.conv2)?;
>         let ys = ys.apply(&self.bn2)?;
>         if let Some(downsample) = &self.downsample {
>             (ys + xs.apply(downsample))?.relu()
>         } else {
>             Ok(ys)
>         }
>     }
> }
> 
> fn basic_layer(
>     vb: VarBuilder,
>     in_planes: usize,
>     out_planes: usize,
>     stride: usize,
>     cnt: usize
> ) -> Result<Sequential<BasicBlock>> {
>     let mut layers = seq();
>     let layer = BasicBlock::new(vb.pp("0"), in_planes, out_planes, stride)?;
>     layers.add(layer);
>     for block_index in 1..cnt {
>         let layer = BasicBlock::new(vb.pp(block_index.to_string()), out_planes, out_planes, 1)?;
>         layers.add(layer);
>     }
>     Ok(layers)
> }
> 
> #[derive(Debug)]
> pub struct ResNet {
>     conv1: Conv2d,
>     bn1: nn::BatchNorm,
>     layer1: Sequential<BasicBlock>,
>     layer2: Sequential<BasicBlock>,
>     layer3: Sequential<BasicBlock>,
>     layer4: Sequential<BasicBlock>,
>     linear: Option<Linear>,
> }
> 
> impl ResNet {
>     pub fn new(
>         vb: VarBuilder,
>         nclasses: Option<usize>,
>         c1: usize,
>         c2: usize,
>         c3: usize,
>         c4: usize
>     ) -> Result<Self> {
>         let conv1 = conv2d(vb.pp("conv1"), 3, 64, 7, 3, 2, false)?;
>         let bn1 = nn::batch_norm(
>             64,
>             BatchNormConfig { eps: 1e-5, remove_mean: false, affine: true },
>             vb.pp("bn1")
>         )?;
>         let layer1 = basic_layer(vb.pp("layer1"), 64, 64, 1, c1)?;
>         let layer2 = basic_layer(vb.pp("layer2"), 64, 128, 2, c2)?;
>         let layer3 = basic_layer(vb.pp("layer3"), 128, 256, 2, c3)?;
>         let layer4 = basic_layer(vb.pp("layer4"), 256, 512, 2, c4)?;
> 
>         let linear = if let Some(n) = nclasses {
>             Some(nn::linear(512, n, vb.pp("fc"))?)
>         } else {
>             None
>         };
> 
>         Ok(Self {
>             conv1,
>             bn1,
>             layer1,
>             layer2,
>             layer3,
>             layer4,
>             linear,
>         })
>     }
> }
> 
> impl Module for ResNet {
>     fn forward(&self, xs: &candle_core::Tensor) -> Result<candle_core::Tensor> {
>         let xs = xs.apply(&self.conv1)?;
>         
>         let xs = xs.apply(&self.bn1)?;
>         let xs = xs.relu()?;
>         //nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
>         let xs = xs.pad_with_zeros(D::Minus2, 1, 1)?;
>         let xs = xs.max_pool2d((3, 2))?;
>         println!("max_pool2d:: {xs}");
>         let xs = xs.apply(&self.layer1)?;
>         // println!("layer1: {xs}");
> 
>         let xs = xs.apply(&self.layer2)?;
>         let xs = xs.apply(&self.layer3)?;
>         let xs = xs.apply(&self.layer4)?;
> 
>         // equivalent to adaptive_avg_pool2d([1, 1]) //avgpool
>         let xs = xs.mean_keepdim(D::Minus2)?.mean_keepdim(D::Minus1)?; //[1, 512, 1, 1]
>         let xs = xs.flatten_from(1)?; //[1,512]
> 
>         match &self.linear {
>             Some(fc) => xs.apply(fc),
>             None => Ok(xs),
>         }
>     }
> }
> 
> fn resnet(
>     vb: VarBuilder,
>     nclasses: Option<usize>,
>     c1: usize,
>     c2: usize,
>     c3: usize,
>     c4: usize
> ) -> Result<ResNet> {
>     ResNet::new(vb, nclasses, c1, c2, c3, c4)
> }
> 
> /// Creates a ResNet-18 model.
> pub fn resnet18(vb: VarBuilder, num_classes: usize) -> Result<ResNet> {
>     resnet(vb, Some(num_classes), 2, 2, 2, 2)
> }
> 
> pub fn resnet18_no_final_layer(vb: VarBuilder) -> Result<ResNet> {
>     resnet(vb, None, 2, 2, 2, 2)
> }
> 
> /// Creates a ResNet-34 model.
> pub fn resnet34(vb: VarBuilder, num_classes: usize) -> Result<ResNet> {
>     resnet(vb, Some(num_classes), 3, 4, 6, 3)
> }
> 
> pub fn resnet34_no_final_layer(vb: VarBuilder) -> Result<ResNet> {
>     resnet(vb, None, 3, 4, 6, 3)
> }
> 
> // Bottleneck versions for ResNet 50, 101, and 152.
> #[derive(Debug)]
> pub struct BottleneckBlock {
>     conv1: Conv2d,
>     bn1: nn::BatchNorm,
>     conv2: Conv2d,
>     bn2: nn::BatchNorm,
>     conv3: Conv2d,
>     bn3: nn::BatchNorm,
>     downsample: Option<Downsample>,
> }
> 
> impl BottleneckBlock {
>     pub fn new(
>         vb: VarBuilder,
>         in_planes: usize,
>         out_planes: usize,
>         stride: usize,
>         e: usize
>     ) -> Result<Self> {
>         let e_dim = e * out_planes;
>         let conv1 = conv2d(vb.pp("conv1"), in_planes, out_planes, 1, 0, 1, false)?;
>         let bn1 = nn::batch_norm(
>             out_planes,
>             BatchNormConfig { eps: 1e-5, remove_mean: false, affine: true },
>             vb.pp("bn1")
>         )?;
>         let conv2 = conv2d(vb.pp("conv2"), out_planes, out_planes, 3, 1, stride, false)?;
>         let bn2 = nn::batch_norm(
>             out_planes,
>             BatchNormConfig { eps: 1e-5, remove_mean: false, affine: true },
>             vb.pp("bn2")
>         )?;
> 
>         let conv3 = conv2d(vb.pp("conv3"), out_planes, out_planes, 1, 0, 1, false)?;
>         let bn3 = nn::batch_norm(
>             out_planes,
>             BatchNormConfig { eps: 1e-5, remove_mean: false, affine: true },
>             vb.pp("bn3")
>         )?;
>         let downsample = downsample(vb.pp("downsample"), in_planes, e_dim, stride)?;
>         Ok(Self {
>             conv1,
>             bn1,
>             conv2,
>             bn2,
>             conv3,
>             bn3,
>             downsample,
>         })
>     }
> }
> 
> impl Module for BottleneckBlock {
>     fn forward(&self, xs: &candle_core::Tensor) -> Result<candle_core::Tensor> {
>         let ys = xs
>             .apply(&self.conv1)?
>             .apply(&self.bn1)?
>             .relu()?
>             .apply(&self.conv2)?
>             .apply(&self.bn2)?
>             .relu()?
>             .apply(&self.conv3)?
>             .apply(&self.bn3)?;
> 
>         if let Some(downsample) = &self.downsample {
>             (xs.apply(downsample) + ys)?.relu()
>         } else {
>             ys.relu()
>         }
>     }
> }
> 
> fn bottleneck_layer(
>     vb: VarBuilder,
>     in_planes: usize,
>     out_planes: usize,
>     stride: usize,
>     cnt: usize
> ) -> Result<Sequential<BottleneckBlock>> {
>     let mut blocks = seq();
>     blocks.add(BottleneckBlock::new(vb.pp("0"), in_planes, out_planes, stride, 4)?);
>     for block_index in 1..cnt {
>         blocks.add(
>             BottleneckBlock::new(
>                 vb.pp(block_index.to_string()),
>                 4 * out_planes,
>                 out_planes,
>                 1,
>                 4
>             )?
>         );
>     }
>     Ok(blocks)
> }
> 
> #[derive(Debug)]
> pub struct BottleneckResnet {
>     conv1: Conv2d,
>     bn1: nn::BatchNorm,
>     layer1: Sequential<BottleneckBlock>,
>     layer2: Sequential<BottleneckBlock>,
>     layer3: Sequential<BottleneckBlock>,
>     layer4: Sequential<BottleneckBlock>,
>     linear: Option<Linear>,
> }
> 
> impl BottleneckResnet {
>     pub fn new(
>         vb: VarBuilder,
>         nclasses: Option<usize>,
>         c1: usize,
>         c2: usize,
>         c3: usize,
>         c4: usize
>     ) -> Result<Self> {
>         let conv1 = conv2d(vb.pp("conv1"), 3, 64, 7, 3, 2, false)?;
>         let bn1 = nn::batch_norm(
>             64,
>             BatchNormConfig { eps: 1e-5, remove_mean: false, affine: true },
>             vb.pp("bn1")
>         )?;
>         let layer1 = bottleneck_layer(vb.pp("layer1"), 64, 64, 1, c1)?;
>         let layer2 = bottleneck_layer(vb.pp("layer2"), 4 * 64, 128, 2, c2)?;
>         let layer3 = bottleneck_layer(vb.pp("layer3"), 4 * 128, 256, 2, c3)?;
>         let layer4 = bottleneck_layer(vb.pp("layer4"), 4 * 256, 512, 2, c4)?;
> 
>         let linear = if let Some(n) = nclasses {
>             Some(nn::linear(4 * 512, n, vb.pp("fc"))?)
>         } else {
>             None
>         };
> 
>         Ok(Self {
>             conv1,
>             bn1,
>             layer1,
>             layer2,
>             layer3,
>             layer4,
>             linear,
>         })
>     }
> }
> 
> impl Module for BottleneckResnet {
>     fn forward(&self, xs: &candle_core::Tensor) -> Result<candle_core::Tensor> {
>         let xs = xs
>             .apply(&self.conv1)?
>             .apply(&self.bn1)?
>             .relu()?
>             .max_pool2d_with_stride(3, 2)?
>             .apply(&self.layer1)?
>             .apply(&self.layer2)?
>             .apply(&self.layer3)?
>             .apply(&self.layer4)?;
>         // equivalent to adaptive_avg_pool2d([1, 1])
>         let xs = xs.mean_keepdim(D::Minus2)?.mean_keepdim(D::Minus1)?;
>         let xs = xs.flatten_to(1)?;
>         match &self.linear {
>             Some(fc) => xs.apply(fc),
>             None => Ok(xs),
>         }
>     }
> }
> 
> fn bottleneck_resnet(
>     vb: VarBuilder,
>     nclasses: Option<usize>,
>     c1: usize,
>     c2: usize,
>     c3: usize,
>     c4: usize
> ) -> Result<BottleneckResnet> {
>     BottleneckResnet::new(vb, nclasses, c1, c2, c3, c4)
> }
> 
> pub fn resnet50(vb: VarBuilder, num_classes: usize) -> Result<BottleneckResnet> {
>     bottleneck_resnet(vb, Some(num_classes), 3, 4, 6, 3)
> }
> 
> pub fn resnet50_no_final_layer(vb: VarBuilder) -> Result<BottleneckResnet> {
>     bottleneck_resnet(vb, None, 3, 4, 6, 3)
> }
> 
> pub fn resnet101(vb: VarBuilder, num_classes: usize) -> Result<BottleneckResnet> {
>     bottleneck_resnet(vb, Some(num_classes), 3, 4, 23, 3)
> }
> 
> pub fn resnet101_no_final_layer(vb: VarBuilder) -> Result<BottleneckResnet> {
>     bottleneck_resnet(vb, None, 3, 4, 23, 3)
> }
> 
> pub fn resnet152(vb: VarBuilder, num_classes: usize) -> Result<BottleneckResnet> {
>     bottleneck_resnet(vb, Some(num_classes), 3, 8, 36, 3)
> }
> 
> pub fn resnet150_no_final_layer(vb: VarBuilder) -> Result<BottleneckResnet> {
>     bottleneck_resnet(vb, None, 3, 8, 36, 3)
> }
> ```
> 
> When I debug using ResNet18, I cannot find the reason for the difference in the tensor of layer1.

That's strange. The results of BasicBlock are different from the official version, and I spent two days carefully comparing the code, but I just can't find the reason.

**Comment by LaurentMazare** [comment]
@kingzcheung could you create a separate issue for this? Seems a bit unrelated to the original thing.

**Comment by LaurentMazare** [comment]
That's amazing, I will add a pointer to it from the main readme to help discoverability. I would also suggest advertising it over reddit r/rust and maybe twitter if you're active there.
Some small nitpicks:
- There are a couple typos, e.g. Embeddin -> Embedding (I can make a PR there though my English is not that great :) ).
- The `linear` and `embedding` helper functions already exist in `candle_nn`, it's good to show how they're implemented as you're doing so that people get a sense of what is going on but probably better to mention the `candle_nn` version so as to avoid the possible cut and paste.
- A bit of the same remark for `LayerNorm`, it also exists in `candle_nn`. Again pretty good to show the implementation, it will help users that want to build a somewhat similar layer but most users that want a standard layer norm should probably use the `candle_nn::{LayerNorm,layer_norm}`.

**Comment by LaurentMazare** [comment]
Closing this as fixed now, thanks again!
